internals/plan/extensions_test.go

```diff
@@ -0,0 +1,680 @@
+			}
+
+			// Verify combined plan YAML.
+			planYAML, err := yaml.Marshal(p)
+			c.Assert(err, IsNil)
+			c.Assert(string(planYAML), Equals, planTest.combinedPlanResultYaml)
+		}
+
+		// Unregister test extensions.
+		for field, _ := range planTest.extensions {
+			plan.UnregisterExtension(field)
+		}
+	}
+}
+
+// writeLayerFiles writes layer files of a test to disk.
+func (s *S) writeLayerFiles(c *C, layersDir string, inputs []*inputLayer) {
+	err := os.MkdirAll(layersDir, 0755)
+	c.Assert(err, IsNil)
+
+	for _, input := range inputs {
+		err := ioutil.WriteFile(filepath.Join(layersDir, fmt.Sprintf("%03d-%s.yaml", input.order, input.label)), reindent(input.yaml), 0644)
+		c.Assert(err, IsNil)
+	}
+}
+
+const xField string = "x-field"
+
+// xExtension implements the LayerSectionExtension interface.
+type xExtension struct{}
+
+func (x xExtension) ParseSection(data yaml.Node) (plan.LayerSection, error) {
+	xs := &xSection{}
+	err := data.Decode(xs)
+	if err != nil {
+		return nil, err
+	}
+	// Propagate the name.
+	for name, entry := range xs.Entries {
+		if entry != nil {
+			xs.Entries[name].Name = name
+		}
+	}
+	return xs, nil
+}
+
+func (x xExtension) CombineSections(sections ...plan.LayerSection) (plan.LayerSection, error) {
+	xs := &xSection{}
+	for _, section := range sections {
+		err := xs.Combine(section)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return xs, nil
+}
+
+func (x xExtension) ValidatePlan(p *plan.Plan) error {
+	var xs *xSection
+	xs = p.Section(xField).(*xSection)
+	if xs != nil {
+		var ys *ySection
+		ys = p.Section(yField).(*ySection)
+
+		// Test dependency: Make sure every Y field in X refer to an existing Y entry.
+		for xEntryField, xEntryValue := range xs.Entries {
+			for _, yReference := range xEntryValue.Y {
+				found := false
+				for yEntryField, _ := range ys.Entries {
+					if yReference == yEntryField {
+						found = true
+						break
+					}
+				}
+				if !found {
+					return fmt.Errorf("cannot find ySection entry %v as required by xSection entry %v ", yReference, xEntryField)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// xSection is the backing type for xExtension.
+type xSection struct {
+	Entries map[string]*X `yaml:",inline,omitempty"`
+}
+
+func (xs *xSection) Validate() error {
+	for field, entry := range xs.Entries {
+		if entry == nil {
+			return fmt.Errorf("cannot have nil entry for %q", field)
+		}
+		// Fictitious test requirement: entry names must start with x
+		if !strings.HasPrefix(field, "x") {
+			return fmt.Errorf("cannot accept entry not starting with letter 'x'")
+		}
+	}
+	return nil
+}
+
+func (xs *xSection) IsZero() bool {
+	return xs.Entries == nil
+}
+
+func (xs *xSection) Combine(other plan.LayerSection) error {
+	otherxSection, ok := other.(*xSection)
+	if !ok {
+		return fmt.Errorf("cannot combine incompatible section type")
+	}
+
+	for field, entry := range otherxSection.Entries {
+		xs.Entries = makeMapIfNil(xs.Entries)
+		switch entry.Override {
```

Comment:
It would be nice to be able to unify the merge/override logic somehow, so each extension didn't have to reimplement it. Any ideas? But maybe that's a pipe dream.

Reply:
You are right that this block repeats itself somewhat with each section implementation.

What I have found, however, while implementing a new extension based on a mature and complex existing schema, is that current Pebble examples where a section is simply a top level ```map[string]*Entry``` structure is not always the case. For example, you could easily have a extension that looks something like this at the top level:

```
type foo struct {
    bar string
    baz map[string]EntryX
}
```

where at the Pebble plan root level, it is positioned as this:

```
sections map[string]section `yaml:",inline"`
```

so that the yaml will look like this:

```
services:
   srv1:
     override:
      :
foo:
   bar: something
   baz:
      zzz1:
        override:
        :
```

So the override levels are not necessarily at the same depth, and the example above have a top level scalar.

Reply:
Resolving.